Objects/dictobject.c

```diff
@@ -6754,19 +6821,52 @@ _PyObject_ManagedDictValidityCheck(PyObject *obj)
 }
 #endif
 
-PyObject *
-_PyObject_GetInstanceAttribute(PyObject *obj, PyDictValues *values,
-                              PyObject *name)
+int
+_PyObject_TryGetInstanceAttribute(PyObject *obj, PyObject *name, PyObject **attr)
 {
     assert(PyUnicode_CheckExact(name));
+    PyDictValues *values = _PyObject_InlineValues(obj);
+    if (!FT_ATOMIC_LOAD_UINT8_RELAXED(values->valid)) {
+        return 1;
+    }
+
     PyDictKeysObject *keys = CACHED_KEYS(Py_TYPE(obj));
     assert(keys != NULL);
     Py_ssize_t ix = _PyDictKeys_StringLookup(keys, name);
     if (ix == DKIX_EMPTY) {
-        return NULL;
+        *attr = NULL;
+        return 0;
+    }
+
+#ifdef Py_GIL_DISABLED
+    PyObject *value = _Py_atomic_load_ptr_relaxed(&values->values[ix]);
+    if (value != NULL && !_Py_TryIncref(&values->values[ix], value)) {
+        int res = 0;
+        // Try again with the object locked...
+        Py_BEGIN_CRITICAL_SECTION(obj);
+        // We could actually race with a dictioanry resize here.  The
+        // inline values will continue to hold references to previous
+        // values, so if we're racing with a resize + insert into the
+        // value we're reading it's just like we won the race.
+        if (!values->valid) {
+            res = 1;
```

Comment:
I don't understand this case:
1) I don't understand the comment in relation to the code
2) It looks like this will cause `_PyObject_GenericGetAttrWithDict` to return `NULL` without an error set

Reply:
This is related to your comment below about this not being thread-safe, i.e. the concurrent modification to the dictionary which could be making the values invalid without `obj` being locked.  I was thinking that even if that happened we'd still have the old values in the concurrently becoming invalid inline values.  But that could also be getting decref'd which would definitely be an issue.

I don't think we'll return NULL w/o an error set?  The return value of 1 here indicates that the caller needs to re-try the lookup from the dictionary as the values are definitely no longer valid.